File: 11-Scheduling-And-Automation/EventOps/executor.py
"""Executor module for running shell commands asynchronously with output logging."""

import logging
import subprocess
import threading

logger = logging.getLogger("eventops.executor")


def execute_sync(action: str, rule_name: str = "") -> subprocess.CompletedProcess:
    """Executes a shell command synchronously and logs stdout/stderr.

    Args:
        action: The shell command line to run.
        rule_name: Optional name of the triggering rule for logging context.

    Returns:
        subprocess.CompletedProcess instance containing execution results.
    """
    tag = f" [{rule_name}]" if rule_name else ""
    logger.info("Executing action%s: %s", tag, action)

    try:
        process = subprocess.run(
            action,
            shell=True,
            capture_output=True,
            text=True,
            check=False,
        )

        if process.stdout and process.stdout.strip():
            for line in process.stdout.strip().splitlines():
                logger.info("[stdout%s] %s", tag, line)

        if process.stderr and process.stderr.strip():
            for line in process.stderr.strip().splitlines():
                logger.warning("[stderr%s] %s", tag, line)

        if process.returncode != 0:
            logger.error(
                "Action%s failed with exit code %d: %s",
                tag,
                process.returncode,
                action,
            )
        else:
            logger.info("Action%s completed successfully.", tag)

        return process

    except (subprocess.SubprocessError, OSError) as e:
        logger.error("Error occurred while executing action%s: %s", tag, e)
        return subprocess.CompletedProcess(
            args=action,
            returncode=-1,
            stdout="",
            stderr=str(e),
        )


def execute_async(action: str, rule_name: str = "") -> threading.Thread:
    """Spawns a background thread to execute a shell command asynchronously.

    Args:
        action: The shell command line to run.
        rule_name: Optional name of the triggering rule for logging context.

    Returns:
        The spawned threading.Thread instance.
    """
    thread = threading.Thread(
        target=execute_sync,
        args=(action, rule_name),
        daemon=True,
        name=f"ExecutorThread-{rule_name or 'daemon'}",
    )
    thread.start()
    return thread


class ActionExecutor:
    """High-level Executor interface for triggering rules."""

    def __init__(self, logger_override: logging.Logger | None = None) -> None:
        self.logger = logger_override or logger

    def run(self, action: str, rule_name: str = "", async_exec: bool = True):
        """Runs the action either asynchronously or synchronously.

        Args:
            action: Shell command to execute.
            rule_name: Name of the rule.
            async_exec: If True, runs in background thread; else runs synchronously.

        Returns:
            threading.Thread if async_exec is True, else subprocess.CompletedProcess.
        """
        if async_exec:
            return execute_async(action, rule_name)
        return execute_sync(action, rule_name)

File: 11-Scheduling-And-Automation/EventOps-Go/pkg/executor/executor.go
// Package executor safely executes shell commands asynchronously or synchronously.
package executor

import (
	"bufio"
	"fmt"
	"io"
	"os/exec"
	"runtime"
	"strings"
	"sync"

	"github.com/Merxxotas/Scripts-DevOps-Practice/11-Scheduling-And-Automation/EventOps-Go/pkg/logger"
)

// Result holds the execution outcome of a shell command.
type Result struct {
	ExitCode int
	Stdout   string
	Stderr   string
	Err      error
}

// ExecuteSync runs a shell command synchronously and logs output.
func ExecuteSync(action string, ruleName string) Result {
	tag := ""
	if ruleName != "" {
		tag = fmt.Sprintf("[%s]", ruleName)
	}

	logger.Info("Executor", "Executing action %s: %s", tag, action)

	var cmd *exec.Cmd
	if runtime.GOOS == "windows" {
		cmd = exec.Command("cmd.exe", "/C", action)
	} else {
		cmd = exec.Command("sh", "-c", action)
	}

	stdoutPipe, err := cmd.StdoutPipe()
	if err != nil {
		logger.Error("Executor", "Failed to create stdout pipe %s: %v", tag, err)
		return Result{ExitCode: -1, Err: err}
	}

	stderrPipe, err := cmd.StderrPipe()
	if err != nil {
		logger.Error("Executor", "Failed to create stderr pipe %s: %v", tag, err)
		return Result{ExitCode: -1, Err: err}
	}

	if err := cmd.Start(); err != nil {
		logger.Error("Executor", "Failed to start command %s: %v", tag, err)
		return Result{ExitCode: -1, Err: err}
	}

	var stdoutBuf, stderrBuf strings.Builder
	var wg sync.WaitGroup
	wg.Add(2)

	// Stream stdout
	go func() {
		defer wg.Done()
		scanAndLog(stdoutPipe, &stdoutBuf, func(line string) {
			logger.Info("Executor", "[stdout %s] %s", tag, line)
		})
	}()

	// Stream stderr
	go func() {
		defer wg.Done()
		scanAndLog(stderrPipe, &stderrBuf, func(line string) {
			logger.Warn("Executor", "[stderr %s] %s", tag, line)
		})
	}()

	err = cmd.Wait()
	wg.Wait() // Ensure all pipe output is read before returning

	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			exitCode = -1
		}
		logger.Error("Executor", "Action %s failed with exit code %d: %s", tag, exitCode, action)
	} else {
		logger.Info("Executor", "Action %s completed successfully.", tag)
	}

	return Result{
		ExitCode: exitCode,
		Stdout:   stdoutBuf.String(),
		Stderr:   stderrBuf.String(),
		Err:      err,
	}
}

// ExecuteAsync spawns a goroutine to execute a shell command asynchronously.
func ExecuteAsync(action string, ruleName string) {
	go func() {
		_ = ExecuteSync(action, ruleName)
	}()
}

func scanAndLog(r io.Reader, buf *strings.Builder, logFn func(string)) {
	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		line := scanner.Text()
		buf.WriteString(line + "\n")
		logFn(line)
	}
}
